File: scripts/git-with-hooks.mjs
#!/usr/bin/env node

/**
 * Git launcher that drops Cursor Source Control's
 * GIT_CONFIG core.hooksPath=/dev/null override so pre-commit runs.
 */

import { spawnSync } from 'node:child_process';
import { existsSync } from 'node:fs';
import { delimiter } from 'node:path';

function isNullDevice(value) {
  if (!value) {
    return false;
  }

  const n = value.trim().toLowerCase().replace(/\//g, '\\');

  return (
    value.trim() === '/dev/null' ||
    n === 'nul' ||
    n === '\\\\.\\nul' ||
    n.endsWith('\\nul')
  );
}

function stripNullDeviceOverrides(env) {
  const count = Number(env.GIT_CONFIG_COUNT ?? 0);

  if (!count) {
    return env;
  }

  const kept = [];

  for (let i = 0; i < count; i += 1) {
    const key = env[`GIT_CONFIG_KEY_${i}`];
    const val = env[`GIT_CONFIG_VALUE_${i}`];
    const interesting =
      /^core\.hooksPath$/i.test(key ?? '') ||
      /^core\.attributesFile$/i.test(key ?? '');

    if (interesting && isNullDevice(val)) {
      continue;
    }

    kept.push([key, val]);
  }

  const next = { ...env };

  for (let i = 0; i < count; i += 1) {
    delete next[`GIT_CONFIG_KEY_${i}`];
    delete next[`GIT_CONFIG_VALUE_${i}`];
  }

  next.GIT_CONFIG_COUNT = String(kept.length);
  kept.forEach(([key, val], i) => {
    next[`GIT_CONFIG_KEY_${i}`] = key;
    next[`GIT_CONFIG_VALUE_${i}`] = val;
  });

  return next;
}

function findGit() {
  if (process.env.PIPELINE_REAL_GIT && existsSync(process.env.PIPELINE_REAL_GIT)) {
    return process.env.PIPELINE_REAL_GIT;
  }

  const name = process.platform === 'win32' ? 'git.exe' : 'git';
  const dirs = (process.env.PATH ?? '').split(delimiter);

  for (const dir of dirs) {
    if (!dir) {
      continue;
    }

    const candidate = `${dir.replace(/[/\\]$/, '')}/${name}`;

    if (!existsSync(candidate)) {
      continue;
    }

    if (/git-with-hooks/i.test(candidate)) {
      continue;
    }

    return candidate;
  }

  return name;
}

const result = spawnSync(findGit(), process.argv.slice(2), {
  env: stripNullDeviceOverrides(process.env),
  stdio: 'inherit',
  windowsHide: true,
});

process.exit(result.status ?? 1);

File: scripts/git-with-hooks.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

/// <summary>
/// Git launcher that drops Cursor Source Control's
/// GIT_CONFIG core.hooksPath=/dev/null override so pre-commit runs.
/// </summary>
internal static class Program
{
    private static int Main(string[] args)
    {
        StripNullDeviceOverrides();

        string git = FindGit();
        if (git == null)
        {
            Console.Error.WriteLine("git-with-hooks: git.exe not found");
            return 1;
        }

        var psi = new ProcessStartInfo();
        psi.FileName = git;
        psi.UseShellExecute = false;
        psi.Arguments = "";

        foreach (string arg in args)
        {
            psi.Arguments += Quote(arg) + " ";
        }

        using (Process child = Process.Start(psi))
        {
            if (child == null)
            {
                Console.Error.WriteLine("git-with-hooks: failed to start git");
                return 1;
            }

            child.WaitForExit();
            return child.ExitCode;
        }
    }

    private static string Quote(string arg)
    {
        if (arg.Length > 0 && arg.IndexOfAny(new[] { ' ', '\t', '"' }) < 0)
        {
            return arg;
        }

        return "\"" + arg.Replace("\"", "\\\"") + "\"";
    }

    private static void StripNullDeviceOverrides()
    {
        string countStr = Environment.GetEnvironmentVariable("GIT_CONFIG_COUNT");
        int count;
        if (string.IsNullOrEmpty(countStr) || !int.TryParse(countStr, out count) || count <= 0)
        {
            return;
        }

        var kept = new List<KeyValuePair<string, string>>();

        for (int i = 0; i < count; i++)
        {
            string key = Environment.GetEnvironmentVariable("GIT_CONFIG_KEY_" + i);
            string val = Environment.GetEnvironmentVariable("GIT_CONFIG_VALUE_" + i);

            if (IsNullDeviceOverride(key, val))
            {
                continue;
            }

            kept.Add(new KeyValuePair<string, string>(key, val));
        }

        for (int i = 0; i < count; i++)
        {
            Environment.SetEnvironmentVariable("GIT_CONFIG_KEY_" + i, null);
            Environment.SetEnvironmentVariable("GIT_CONFIG_VALUE_" + i, null);
        }

        Environment.SetEnvironmentVariable("GIT_CONFIG_COUNT", kept.Count.ToString());

        for (int i = 0; i < kept.Count; i++)
        {
            Environment.SetEnvironmentVariable("GIT_CONFIG_KEY_" + i, kept[i].Key);
            Environment.SetEnvironmentVariable("GIT_CONFIG_VALUE_" + i, kept[i].Value);
        }
    }

    private static bool IsNullDeviceOverride(string key, string val)
    {
        if (string.IsNullOrEmpty(key) || string.IsNullOrEmpty(val))
        {
            return false;
        }

        bool interesting =
            key.Equals("core.hooksPath", StringComparison.OrdinalIgnoreCase)
            || key.Equals("core.attributesFile", StringComparison.OrdinalIgnoreCase);

        if (!interesting)
        {
            return false;
        }

        string n = val.Trim();

        return n.Equals("/dev/null", StringComparison.OrdinalIgnoreCase)
            || n.Equals("nul", StringComparison.OrdinalIgnoreCase)
            || n.Equals("\\\\.\\nul", StringComparison.OrdinalIgnoreCase)
            || n.Equals("\\Device\\Null", StringComparison.OrdinalIgnoreCase);
    }

    private static string FindGit()
    {
        string fromEnv = Environment.GetEnvironmentVariable("PIPELINE_REAL_GIT");

        if (!string.IsNullOrEmpty(fromEnv) && File.Exists(fromEnv))
        {
            return fromEnv;
        }

        string[] wellKnown = new[]
        {
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles),
                "Git",
                "cmd",
                "git.exe"),
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
                "Git",
                "cmd",
                "git.exe"),
        };

        foreach (string candidate in wellKnown)
        {
            if (File.Exists(candidate))
            {
                return candidate;
            }
        }

        string pathEnv = Environment.GetEnvironmentVariable("PATH") ?? "";

        foreach (string dir in pathEnv.Split(Path.PathSeparator))
        {
            if (string.IsNullOrEmpty(dir))
            {
                continue;
            }

            string candidate = Path.Combine(dir, "git.exe");

            if (!File.Exists(candidate))
            {
                continue;
            }

            if (candidate.IndexOf("git-with-hooks", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                continue;
            }

            return candidate;
        }

        return null;
    }
}
